README.md
<p align="center"><a href="https://galantelab.github.io/sandy/"><img src="share/imgs/sandy.jpeg" alt="sandy logo"></a></p>
<h2 align="center">A straightforward and complete next-generation sequencing read simulator</h2>

<p align="center">
  <a href="https://github.com/galantelab/sandy/actions/workflows/ci.yml"><img alt="" src="https://github.com/galantelab/sandy/actions/workflows/ci.yml/badge.svg?branch=dev" align="center"></a>
  <a href="https://badge.fury.io/pl/App-Sandy"><img alt="" src="https://badge.fury.io/pl/App-Sandy.svg" align="center"></a>
  <a href="https://hub.docker.com/r/galantelab/sandy/tags"><img alt="" src="https://img.shields.io/badge/docker-%230db7ed.svg?style=flat&logo=docker&logoColor=white" align="center"></a>
</p>

**Sandy** is a bioinformatics tool that provides a simple engine to simulate next-generation
sequencing (NGS) reads for genomic and transcriptomic pipelines. Simulated data works as
experimental control \- a key step to optimize NGS analysis - in comparison to hypothetical
models. **Sandy** is a straightforward, easy-to-use, fast and highly customizable tool that
generates reads requiring only a fasta file as input. **Sandy** can simulate **single-end**
and **paired-end** reads from both DNA and RNA sequencing as if produced from the most used
second and third-generation platforms. The tool also tracks a built-in database with predefined
models extracted from real data for sequencer **quality-profiles** (i.e.
[Illumina](https://www.illumina.com/) *hiseq*, *miseq*, *nextseq*), **expression-matrices**
generated from [GTExV8](https://www.gtexportal.org/home/) data for 54 human tissues, and
**genomic-variations** such as SNVs and Indels from [1KGP](https://www.internationalgenome.org/)
and gene fusions from [COSMIC](https://cancer.sanger.ac.uk/cosmic).

For full documentation, please visit <https://galantelab.github.io/sandy/>.

## Features

* Simulate DNA and RNA sequencing

    Simulate **single-end** (long and short fragments) and **paired-end** sequencing reads for
    **genome** and **transcriptome** analysis. The simulation can be customized with raffle seed,
    sequencing coverage, number of reads, fragment mean, output formats (`fastq`, `sam` and their
    compressed versions `fastq.gz` and `bam`), sequence identifier (header of entries in `fastq`)
    and much more.

* Sequencer **quality-profile**

    **Sandy** generates `fastq` quality entries that mimic the [Illumina](https://www.illumina.com/),
    [PacBio](https://www.pacb.com/) and [Nanopore](https://nanoporetech.com/) sequencers, as well as
    generating the *phred-score* using a statistical model based on the *poisson* distribution.

* RNA-Seq **expression-matrix**

    It is possible to simulate a RNA-Seq which reflects the abundance of gene expression for transcripts
    and genes of a given tissue. For this purpose, **expression-matrices** were created from the gene
    expression data of 54 tissues of the [GTExV8](https://www.gtexportal.org/home/) project.

* Whole-genome sequencing with **genomic-variiation**

    The user can tune the reference genome (eg [GRCh38.p13.genome.fa.gz](https://www.gencodegenes.org/human/)),
    adding homozygous or heterozygous **genomic-variations** such as SNVs, Indels, gene fusions and other
    types of structural variations (eg CNVs, retroCNVs). **Sandy** has in its database **genomic-variations**
    obtained from the [1KGP](https://www.internationalgenome.org/) and from
    [COSMIC](https://cancer.sanger.ac.uk/cosmic).

* Custom user models

    Users can include their models for **quality-profile**, **expression-matrix** and **genomic-variation**
    in order to adapt the simulation to their needs.

* Custom sequence identifier

    The sequence identifier, as the name implies, is a string that identifies a biological sequence (usually
    nucleotides) within a sequencing data. For example, the `fasta` format includes the sequence identifier
    always after the `>` character at the beginning of the line; the `fastq` format always includes it after
    the `@` character at the beginning of the line; the `sam` format uses the first column (called the
    *query template name*).

    | Sequence identifier | File format |
    | :-- | :-: |
    | \>**MYID and Optional information**<br />ATCGATCG | `fasta` |
    | @**MYID and Optional information**<br />ATCGATCG<br />+<br />ABCDEFGH | `fastq` |
    | **MYID** 99 chr1 123456 20 8M chr1 123478 30 ATCGATCG ABCDEFGH | `sam` |

    Sequence identifiers may be customized in output using a format string passed by the user. This format
    is a combination of literal and escaped characters, in a similar fashion to that used in C programming
    language’s `printf` function.

    For example, simulating a paired-end sequencing you can add the read length, read position and mate
    position into all sequence identifiers with the following format:

        %i.%U read=%c:%t-%n mate=%c:%T-%N length=%r

    In this case, results in `fastq` format would be:

        ==> Into R1
        @SR.1 read=chr6:979-880 mate=chr6:736-835 length=100
        ...
        ==> Into R2
        @SR.1 read=chr6:736-835 mate=chr6:979-880 length=100

## Installation

There are two recommended ways to obtain **Sandy**: Pulling the official [Docker](https://www.docker.com/)
image and installing through [CPAN](https://metacpan.org/).

### Docker

Assuming that `docker` is already installed on your server, simply run the command:

    $ docker pull galantelab/sandy

For more details, see [docker/README.md](https://github.com/galantelab/sandy/blob/master/docker/README.md)
file.

### CPAN

#### Prerequisites

Along with `perl`, you must have `zlib`, `gcc`, `make` and `cpanm` packages installed:

- Debian/Ubuntu

        % apt-get install perl zlib1g-dev gcc make cpanminus

- CentOS/Fedora

        % yum install perl zlib gcc make perl-App-cpanminus

- Archlinux

        % pacman -S perl zlib gcc make cpanminus

#### Installing with `cpanm`

Install **Sandy** with the following command:

    % cpanm App::Sandy

If you concern about speed, you can avoid testing with the flag `--notest`:

    % cpanm --notest App::Sandy

For more details, see [INSTALL](https://github.com/galantelab/sandy/blob/master/INSTALL) file

## Acknowledgments

| Institution | Site |
| :-- | :-: |
| Coordination for the Improvement of Higher Level Personnel | [CAPES](http://www.capes.gov.br/) |
| The São Paulo Research Foundation | [FAPESP](https://fapesp.br/en/about) |
| Teaching and Research Institute from Sírio-Libanês Hospital | [Galantelab](https://www.bioinfo.mochsl.org.br/) |

## License

This is free software, licensed under:

    The GNU General Public License, Version 3, June 2007



Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.25      2023-09-19 14:55:24-03:00 America/Sao_Paulo

  - Remove version from transcript_id, example:
    ENST00006666.9 will become ENST00006666

0.24      2023-07-31 12:53:42-03:00 America/Sao_Paulo

  - When running 'sandy transcriptome' save transcript counts
    and gene counts into separated files
  - Use GNU/GSL RNG and Randist for randlux algorithm
  - Update db.sqlite3 with GTEXv8 expression patterns
  - Enable persistent storage of sandy db in docker
  - Uppercase all FASTA bases
  - Apply NNN blacklist filter

0.23      2019-03-10 17:20:19-03:00 America/Sao_Paulo

  - Fix database permissions
  - Improve README.md and docker/README.md

0.22      2018-12-20 16:18:05-02:00 America/Sao_Paulo

  - Change docker repository to the namespace galantelab
  - Change genome command option '--structural-variation'
    to '--genomic-variation'

0.21      2018-09-18 20:07:39-03:00 America/Sao_Paulo

  - Set expression-matrix as optional
  - Fix bash/zsh completion installation
  - Distinguish count file between genome (coverage)
    and transcriptome (abundance)
  - Add compression level option
  - Set minimum perl version to 5.014

0.20      2018-09-14 01:49:09-03:00 America/Sao_Paulo

  - Add .travis.yml
  - Fix count file names
  - Testing compression level option

0.19      2018-08-10 21:10:27-03:00 America/Sao_Paulo

  - Implement structural variation
  - Add variation database
  - Add bash/zsh completions

0.18      2018-05-06 02:36:33-03:00 America/Sao_Paulo

  - Add citation command

0.17      2018-05-05 22:44:51-03:00 America/Sao_Paulo

  - Fix bug in fasta_tree and fasta_rtree
  - Populate quality and expression databases
  - Improve documentation
  - Rename project from App-SimulateReads to App-Sandy

0.16      2018-04-16 00:59:15-03:00 America/Sao_Paulo

  - Update quality profile database in order to keep it simple
  - Turn read-size option settable only when quality-profile="poisson".
    Otherwise, if quality-profile comes from database, it will override
    the default/user-provided value

0.15      2018-04-15 02:15:43-03:00 America/Sao_Paulo

  - Substitute expression-matrix for weight-file
  - Add an expression-matrix database to provide vendor
    options for the users
  - Change main commands to: genome, transcriptome, custom,
    quality and expression. The first three manage simulations
    and the last two manage databases

0.14      2018-03-29 03:29:43-03:00 America/Sao_Paulo

  - Add seed option for reproducible simulations
  - Change weighted raffle from 'role' to 'class'

0.13      2018-03-25 20:40:24-03:00 America/Sao_Paulo

  - Add save counts
  - SeqId is now case sensitive

0.12      2018-03-24 13:38:37-03:00 America/Sao_Paulo

  - Add sequence identifier template

0.11      2018-03-23 18:44:59-03:00 America/Sao_Paulo

  - Fix bug when weight-file has just one entry and it is equal to "1"
  - Split digest command into genome, transcriptome and custom

0.10      2017-12-03 04:27:18-02:00 America/Sao_Paulo

  - Add version option

0.09      2017-11-28 23:24:58-02:00 America/Sao_Paulo

  - Fix bug concerning perl versions (5.010 and >=5.018) and
    experimental::smartmatch warnings

0.08      2017-11-27 22:21:06-02:00 America/Sao_Paulo

  - Fix indexed weight file. If an entry does not exist at the indexed
    fasta, then it is removed from the weight file hash

0.07      2017-11-27 11:55:16-02:00 America/Sao_Paulo

  - App no more dies if an indexed weighted file has not a valid seqid
    recorded at a validated indexed fasta
  - Add Dockerfile for the project

0.06      2017-10-29 04:28:50-02:00 America/Sao_Paulo

  - Corrected My::Types to App::SimulateReads::Types in Base class
  - App no more dies if an indexed fasta sequence has not passed the
    length constraints. Instead, that sequence are deleted from the hash

0.05      2017-10-02 22:57:50-03:00 America/Sao_Paulo

  - Remove Statistics::Basic
  - Test::Perl::Critic and Test::UseAllModules were moved to [TestRecommends]

0.04      2017-09-25 19:16:51-03:00 America/Sao_Paulo

  - Minimum perl version set to v5.10

0.03      2017-09-17 22:50:47-03:00 America/Sao_Paulo

  - Split App::SimulateReads by creating a CLI::App, CLI::Command interface

0.02      2017-09-16 02:18:05-03:00 America/Sao_Paulo

  - Fix README.md issue with PodWeaver by using ReadmeAnyFromPod

0.01      2017-09-15 23:03:14-03:00 America/Sao_Paulo

  - Initial version


lib/My/Types.pm
#
#===============================================================================
#
#         FILE: Types.pm
#
#  DESCRIPTION: Type constraints
#
#        FILES: ---
#         BUGS: ---
#        NOTES: ---
#       AUTHOR: Thiago Miller
# ORGANIZATION: IEP - Hospital Sírio-Libanês
#      VERSION: 1.0
#      CREATED: 19-05-2017 21:00:05
#     REVISION: ---
#===============================================================================

package My::Types;
 
use Moose::Util::TypeConstraints;

subtype 'My:IntGt0'
	=> as      'Int'
	=> where   { $_ > 0 } 
	=> message { "Value must be an integer greater than zero, not '$_'" };

subtype 'My:IntGe0'
	=> as      'Int'
	=> where   { $_ >= 0 } 
	=> message { "Value must be an integer greater or equal to zero, not '$_'" };

subtype 'My:NumGt0'
	=> as      'Num'
	=> where   { $_ > 0 } 
	=> message { "Value must be a number greater than zero, not '$_'" };

subtype 'My:NumGe0'
	=> as      'Num'
	=> where   { $_ >= 0 } 
	=> message { "Value must be a number greater or equal to zero, not '$_'" };

subtype 'My:NumHS'
	=> as      'Num'
	=> where   { $_ >= 0 && $_ <= 1 }
	=> message { "Value must be a number between zero and one, not '$_'" };

subtype 'My:File'
	=> as      'Str'
	=> where   { -f $_ }
	=> message { "'$_' must be a file" };

subtype 'My:Fasta'
	=> as      'My:File'
	=> where   { $_ =~ /.+\.(fasta|fa|fna|ffn)(\.gz)*$/ }
	=> message { "'$_' must be a fasta file: Check the extension (.fasta, .fa, .fna, .ffn - compressed, or not, by gzip, as in .fasta.gz etc)" };

subtype 'My:Weight'
	=> as      'HashRef'
	=> where   { exists $_->{down} && exists $_->{up} && exists $_->{feature} }
	=> message { "'$_' is not a Weight object" };

subtype 'My:Weights'
	=> as      'ArrayRef[My:Weight]'
	=> message { "'$_' is not a Weight object array" };

subtype 'My:QualityP'
	=> as      'Str';

coerce 'My:QualityP'
	=> from    'Str'
	=> via     { lc $_ };

subtype 'My:QualityH'
	=> as      'HashRef'
	=> where   { exists $_->{matrix} && exists $_->{deepth} }
	=> message { "'$_' is not a valid quality hash" };

subtype 'My:IdFa'
	=> as      'HashRef'
	=> where   { exists $_->{seq} && exists $_->{size} }
	=> message { "'$_' is not a valid fasta id" };

subtype 'My:IdxFasta'
	=> as      'HashRef[My:IdFa]'
	=> message { "'$_' is not a valid indexed fasta" };

subtype 'My:StrandBias'
	=> as      'Str'
	=> where   { $_ eq 'plus' || $_ eq 'minus' || $_ eq 'random' }
	=> message { "'$_' is not a valid strand-bias: 'plus', 'minus' or 'random'" };

subtype 'My:SeqIdWeight'
	=> as      'Str'
	=> where   { $_ eq 'length' || $_ eq 'same' || $_ eq 'file' }
	=> message { "'$_' is not a valid seqid-weight: 'length', 'same' or 'file'" };

subtype 'My:SeqType'
	=> as      'Str'
	=> where   { $_ eq 'single-end' || $_ eq 'paired-end' }
	=> message { "'$_' is not a valid sequencing-type: 'single-end' or 'paired-end'" };

subtype 'My:CountLoopBy'
	=> as      'Str'
	=> where   { $_ eq 'coverage' || $_ eq 'number-of-reads' }
	=> message { "'$_' is not a valid count_loops_by: 'coverage' or 'number-of-reads'" };

1; ## --- end class My::Types
